omnibus/files/private-chef-ctl-commands/cleanup_bifrost.rb

```diff
@@ -0,0 +1,231 @@
+  elsif options[:batch_size] <= 0
+    puts "Invalid batch size: #{options[:batch_size]}"
+    exit(1)
+  end
+
+  # NOTE(ssd) 2017-07-24:
+  # For the most part, the tracking table manages in-flight requests for us.
+  # However, it is still possible that the following sequence happens:
+  #    - new authz_id in bifrost database
+  #    - tracking table inserted
+  #    - known_actor_list fetched
+  #    - new record in oc-chef database
+  #
+  # In the this case, the authz_id won't be in our list of known
+  # clients, won't be in our tracking table, but will show up in our
+  # search for orphaned authz_ids.
+  #
+  # The wait_time is a safety-timer we use to avoid this race.  The
+  # default of 50 seconds is:
+  #
+  #   (# of db/upstream calls after auth_actor creation * 5 seconds) * 2
+  #
+  if !options[:wait_time]
+    options[:wait_time] = 50
+  elsif options[:wait_time] <= 0
+    puts "Invalid wait time: #{options[:wait_time]}"
+    exit(1)
+  end
+
+  if options[:cleanup_only]
+    remove_bifrost_tracking_table(bifrost_db)
+    exit(0)
+  end
+
+  if options[:estimate_only]
+    print_and_return_estimate(known_actor_list, bifrost_db)
+    exit(0)
+  end
+
+  run_cleanup(bifrost_db, options[:batch_size], options[:wait_time])
+end
+
+def run_cleanup(bifrost_db, batch_size, wait_time)
+  safety_check(bifrost_db)
+  install_bifrost_tracking_table(bifrost_db)
+
+  begin
+    puts "Sleeping #{wait_time} seconds to account for in-flight requests not captured by tracking table"
+    sleep wait_time
+    estimated_deletion_count = print_and_return_estimate(known_actor_list, bifrost_db)
+
+    if estimated_deletion_count <= 0
+      puts "Estimated deletion count 0. Aborting"
+      exit(0)
+    end
+
+    run_bifrost_scan(known_actor_list, batch_size, bifrost_db)
+  ensure
+    remove_bifrost_tracking_table(bifrost_db)
+  end
+end
+
+def known_actor_list
+  @known_actor_list ||= timed "Fetching initial opscode_chef clients and user list" do
+    clients_and_users = erchef_db.exec("SELECT authz_id FROM clients UNION select authz_id FROM users")
+    clients_and_users.map do |real_actor|
+      real_actor['authz_id']
+    end
+  end
+end
+
+def safety_check(db)
+  res = db.exec("SELECT * FROM pg_tables
+                          WHERE tablename='cleanup_tracking_auth_actors'")
+  if res.ntuples > 0
+    puts "ERROR: cleanup_tracking_auth_actors already exists.  cleanup-bifrost may be running."
+    puts "ERROR: If you are sure cleanup-bifrost is not running, you can clean up the tracking tables with: chef-backend-ctl cleanup-bifrost --force-cleanup"
+    exit(1)
+  end
+end
+
+def install_bifrost_tracking_table(db)
+  timed "Installing tracking tables into bifrost database" do
+    db.exec(CREATE_SQL)
+  end
+end
+
+def install_known_actor_table(db, list)
+  list_str = "('" + list.join("'), ('") + "')"
+  timed "Populating known actor table" do
+    db.exec(CREATE_SQL2)
+    db.exec("INSERT INTO cleanup_known_auth_actors VALUES #{list_str}")
```

Comment:
Yes. I can see two situations (1) the list is so large that holding it in memory and doing that join is too expensive or (2) the list gets so long that the INSERT query grinds to a halt.  

For some reason I was under the impression that our version of the PG gem didn't support COPY, but it appears that it does just fine from what I can see in git, so one option would be to use `COPY`. Another would be to chuck the list and do multiple INSERTS.